src/sync/subscribable.hpp

```diff
@@ -0,0 +1,104 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_SUBSCRIBABLE_HPP
+#define REALM_SUBSCRIBABLE_HPP
+
+#include <cstdint>
+#include <unordered_map>
+
+namespace realm {
+
+/// Generic subscribable that allows for coarse, manual notifications
+/// from class type T.
+template <class T>
+struct Subscribable {
+    /// Token that identifies an observer.
+    /// Unsubscribes when deconstructed to
+    /// avoid dangling observers.
+    struct Token {
+        Token(Subscribable& subscribable, uint64_t token)
+        : m_subscribable(subscribable)
+        , m_token(std::make_unique<uint64_t>(token))
+        {
+        }
+        Token(Token&& other)
+        : m_subscribable(std::move(other.m_subscribable))
+        , m_token(std::move(other.m_token))
+        {
+        }
+        Token& operator=(Token&& other) {
+            m_subscribable = std::move(other.m_subscribable);
+            m_token = std::move(other.m_token);
+        }
+        Token(const Token&) = delete;
+        Token& operator=(const Token&) = delete;
+
+        ~Token()
+        {
+            if (m_token) {
+                m_subscribable.get().unsubscribe(*this);
+            }
+        }
+
+        uint64_t value() const
+        {
+            return *m_token;
+        }
+    private:
+        std::reference_wrapper<Subscribable> m_subscribable;
+        std::unique_ptr<uint64_t> m_token;
```

Comment:
This could just be a `uint64_t` and `m_subscribable` could be a pointer that's set to null when it's moved from.

Reply:
I was under the impression that we tried to avoid raw pointers, but yes, I agree, this seems like applicable an applicable case.

Reply:
We avoid *owning* raw pointers. Pointers for non-owning nullable references are totally fine.